Добавлена лабораторная работа по trie деревьям

Shardenn opened the pull request.

Shardenn committed:
Test adding

Added main.cpp to test

Shardenn committed:
Uploaded trie lab

tapoton commented on Vikarchuk/trie/trie/main.cpp:
```diff
@@ -0,0 +1,110 @@
+	void init(Node *node)
+	{
+		node->prefix_count = 0;
+		node->is_end = true;
```
здесь точно true? у вас false нигде не выставляется, по всему дереву true

tapoton commented on Vikarchuk/trie/trie/main.cpp:
```diff
@@ -0,0 +1,110 @@
+	{
+		node->prefix_count = 0;
+		node->is_end = true;
+		for (unsigned int i = 0; i <= 26; i++)
```
`<= 26`:
1. Зачем вам `ALPHABET_SIZE`, если вы все время используете 26?
2. от 0 до 26 включительно у вас всего 27 символов, один лишний.

Shardenn committed:
Исправления в связи с комментариями

Исправлена 34 строка: нестрогий знак на строгий, 26 на константу.
54 строка: если создаётся следующий узел за текущим, is_end текущего
становится false.

tapoton commented on Vikarchuk/trie/trie/main.cpp:
```diff
@@ -0,0 +1,111 @@
+		head = new Node;
+		head->prefix_count = 0;
+		head->is_end = false;
+		for (unsigned int i = 0; i <= 26; i++)
```
а тут?

tapoton commented on Vikarchuk/trie/trie/main.cpp:
```diff
@@ -0,0 +1,111 @@
+				init(current->child[letter]);
+			}
+			current->child[letter]->prefix_count++;
+			current->is_end = false;
```
если у меня было добавлено слово qwer и я добавлю qwerty, то для узла 'r' флаг будет выставлен в false, слово будет утеряно, я прав?

Shardenn replied:
Я ведь сейчас отправил примерно 6 одинаковых запросов? Или всё в порядке?
Программу сейчас исправлю.

Shardenn committed:
true в init() изменено на false

В конструкторе стояло false по умолчанию, в init() забыл поменять на то
же. Теперь каждая буква имеет дефолтно false в is_end. При добавдении же
слова последняя буква получает true в is_end.
